File: 2. Structural Design Patterns/Proxy/c++/ProxyExample.cpp
#include <iostream>
#include <string>
using namespace std;

// Subject interface
class Image {
public:
    virtual void display() = 0;
    virtual ~Image() {}
};

// Real Subject
class RealImage : public Image {
private:
    string filename;

    void loadFromDisk() {
        cout << "Loading image from disk: " << filename << endl;
    }

public:
    RealImage(const string& file) : filename(file) {
        loadFromDisk();
    }

    void display() override {
        cout << "Displaying image: " << filename << endl;
    }
};

// Proxy
class ProxyImage : public Image {
private:
    string filename;
    RealImage* realImage;

public:
    ProxyImage(const string& file) : filename(file), realImage(nullptr) {}

    ~ProxyImage() {
        delete realImage;
    }

    void display() override {
        if (!realImage) {
            realImage = new RealImage(filename); // Lazy loading
        }
        realImage->display();
    }
};
int main() {
    Image* image = new ProxyImage("high_res_photo.png");

    cout << "Image created, but not loaded yet.\n";

    cout << "\nFirst display call:\n";
    image->display();  // Will load from disk and display

    cout << "\nSecond display call:\n";
    image->display();  // Will just display (no reloading)

    delete image;
    return 0;
}

File: 2. Structural Design Patterns/Proxy/Python/ProxyExample.py
# Proxy Pattern Example in Python
# Image viewer with lazy-loading proxy

from abc import ABC, abstractmethod

class Image(ABC):
    @abstractmethod
    def display(self):
        pass

class RealImage(Image):
    def __init__(self, filename):
        self.filename = filename
        self._load_from_disk()
    
    def _load_from_disk(self):
        print(f"Loading image: {self.filename}")
    
    def display(self):
        return f"Displaying: {self.filename}"

class ProxyImage(Image):
    def __init__(self, filename):
        self.filename = filename
        self.real_image = None
    
    def display(self):
        if self.real_image is None:
            self.real_image = RealImage(self.filename)
        return self.real_image.display()

def main():
    image = ProxyImage("photo.jpg")
    print(image.display())  # loads then displays
    print(image.display())  # displays without loading

if __name__ == "__main__":
    main()
